library/std/src/time.rs

```diff
@@ -125,6 +126,25 @@ pub use core::time::FromSecsError;
 /// > structure cannot represent the new point in time.
 ///
 /// [`add`]: Instant::add
+///
+/// ## Monotonicity
+///
+/// All [tier 1] and most tier 2 platforms with std support rely on monotonicity guarantees
```

Comment:
This sort of misleadingly (I think) implies that tier 3 platforms are *not* relying on these guarantees. I think what we're trying to say here is that "Instant depends on the underlying platform to provide a monotonic clock" [and sometimes that platform lies/gets it wrong].

Can we avoid talking about tiers at all, perhaps? It doesn't seem particularly meaningful, given we "need" the same guarantees realistically regardless of tier.

Reply:
Some of the lower tiers just don't support `Instant` at all and return errors.

Reply:
I presume we always either return errors or ask for a monotonic clock? Not *all* tier 3 targets are not-supporting of instant, right?

Reply:
This SOLID (tier3 platform) comment reads like `Instant` is using a system timer rather than a cycle counter.

https://github.com/rust-lang/rust/blob/72e74d7b9cf1a7901650227e74650f1fcc797600/library/std/src/sys/itron/time.rs#L21-L24

Reply:
ping @Mark-Simulacrum

Do you still want me to update the wording? The documentation can't speak for all (especially 3rd-party) or future platforms. Some OSes might not guarantee monotonicity but still have a high resolution time source. The question is whether saturating will be enough in those cases or we should keep the mutex/atomics solution around for those cases?

Reply:
Hm... Not sure. I missed your comment before, but it seems to me that if we know Instant isn't going to be monotonic then saturating can yield fairly unexpected and (in some sense) wrong behavior.

I'm not sure either whether the atomic/mutex solution really helps there - if we're not using a source that even claims to be monotonic, we can't really effectively fake it atop imo, since you'd just get a long period of always zero or panics.

I think we may need some team discussion, but my initial inclination is that it may be worth cfg'ing out and/or deprecating Instant for platforms we do not know to have an API that claims monotonicity. (Even if in some hardware or virtualization environments that monotonicity is buggy).

Reply:
DIscussed in libs-api meeting -- I think we probably want to rephrase to something like this:

On tier 1 platforms, Instant will request a monotonic time from the underlying system. On other platforms, the standard library makes a best-effort attempt to use a monotonic API. Note that regardless of the target, the underlying platform may suffer from bugs which break the monotonic guarantee [...].

The trailer here is the rest of the paragraph as typical.

@the8472 let me know if that makes sense to you or you think we should otherwise rephrase.

Reply:
Yeah, I specifically wanted to avoid a promise that sounded like we *definitely* use the API (e.g., we may not on a tier 2/3 platform because no one has told us it exists yet). I'm not sure how best to communicate that; my best-effort was an attempt to do so.